docs/DeveloperGuide.adoc

```diff
@@ -218,29 +218,53 @@ image::CommitActivityDiagram.png[]
+** Cons: We must ensure that the implementation of each individual command is correct.
 
 ===== Aspect: Data structure to support the undo/redo commands
 
-* **Alternative 1 (current choice):** Use a list to store the history of address book states.
+* **Alternative 1 (current choice):** Use a list to store the history of CardiBuddy states.
 ** Pros: Easy for new Computer Science student undergraduates to understand, who are likely to be the new incoming developers of our project.
-** Cons: Logic is duplicated twice. For example, when a new command is executed, we must remember to update both `HistoryManager` and `VersionedAddressBook`.
+** Cons: Logic is duplicated twice. For example, when a new command is executed, we must remember to update both `HistoryManager` and `VersionedCardiBuddy`.
 * **Alternative 2:** Use `HistoryManager` for undo/redo
 ** Pros: We do not need to maintain a separate list, and just reuse what is already in the codebase.
 ** Cons: Requires dealing with commands that have already been undone: We must remember to skip these commands. Violates Single Responsibility Principle and Separation of Concerns as `HistoryManager` now needs to do two different things.
 // end::undoredo[]
 
-// tag::dataencryption[]
-=== [Proposed] Data Encryption
+=== Search and Filter feature
+Users can choose to find a `Deck` or a `Flashcard` through:
 
-_{Explain here how the data encryption feature will be implemented}_
+* searching by keywords from the title of a `Deck` or from the `Question` or `Answer` from a `Flashcard`.
+* filtering by tags from the `Tag(s)` associated with a `Deck` or a `Flashcard`
 
-// end::dataencryption[]
+Both `search` and `filter` have the and-or functionality where adding `&` between keywords returns decks and flashcards
+with both terms before and after the `&` symbol.
+For example, `filter Y1 & hard` will return decks and flashcards that are tagged with both `Y1` and `hard`.
+
+The following sequence diagram shows how the `Search` command works:
+
+image::SearchFilterDiagram.png[]
+
+Figure X. Component interactions for a Search command
+
+Calling the `search` command creates a predicate based on the given keywords and passed as an argument to a `SearchCommand` object.
+The `execute()` method is then called to filter the decks accordingly to the predicate and it returns a `CommandResult` object.
+
+Future implementations would return flashcards and decks that are filtered accordingly to the predicate. (by v1.3)
+
+=== Test feature
+Users can start a <<test-session>> with a chosen `Deck` by:
+
+* The `index` of the deck will be provided byt he user and parsed by Cardi Buddy. A `TestSession` object containing the indicated `Deck` is created.
+* The `TestSession` object updates Cardibuddy's `UniqueFlashcardList` to display the first flashcard in the queue, along with the other flashcards to be tested.
+
+The following is a proposed diagram of how the `test` feature should work:
+
+image::TestSequenceDiagram.png[]
```

Comment:
Under "d:TestCommand", an object d seems to be created but object 't' is returned instead. 
Also, the figure is not labelled.